examples/langchain-rag/app.py

```diff
@@ -0,0 +1,160 @@
+from langchain.chat_models import AzureChatOpenAI
+from langchain.document_loaders import PyPDFLoader, WebBaseLoader
+from langchain.embeddings import CacheBackedEmbeddings, OpenAIEmbeddings
+from langchain.schema.output import ChatGenerationChunk, GenerationChunk
+from langchain.storage import LocalFileStore
+from langchain.text_splitter import RecursiveCharacterTextSplitter
+from langchain.vectorstores import Chroma
+
+from aidial_sdk import DIALApp
+from aidial_sdk.chat_completion import ChatCompletion, Choice, Request, Response
+
+DIAL_URL = os.getenv("DIAL_URL")
+
+text_splitter = RecursiveCharacterTextSplitter.from_tiktoken_encoder(
+    chunk_size=256, chunk_overlap=0
+)
+
+embedding_store = LocalFileStore("./cache/")
+
+
+class CustomCallbackHandler(AsyncCallbackHandler):
+    def __init__(self, choice: Choice):
+        self._choice = choice
+
+    async def on_chat_model_start(
+        self,
+        serialized: Dict[str, Any],
+        messages: List[List[Any]],
+        *,
+        run_id: UUID,
+        parent_run_id: Optional[UUID] = None,
+        tags: Optional[List[str]] = None,
+        metadata: Optional[Dict[str, Any]] = None,
+        **kwargs: Any,
+    ) -> Any:
+        pass
+
+    async def on_llm_new_token(
+        self,
+        token: str,
+        *,
+        chunk: Optional[Union[GenerationChunk, ChatGenerationChunk]] = None,
+        run_id: UUID,
+        parent_run_id: Optional[UUID] = None,
+        tags: Optional[List[str]] = None,
+        **kwargs: Any,
+    ) -> None:
+        self._choice.append_content(token)
+
+
+class SimpleRAGApplication(ChatCompletion):
+    async def chat_completion(
+        self, request: Request, response: Response
+    ) -> None:
+        collection_name = str(uuid4())
+
+        with response.create_single_choice() as choice:
+            first_user_message = request.messages[0].content or ""
+            if first_user_message.endswith(".pdf"):
+                loader = PyPDFLoader(first_user_message)
+            else:
+                loader = WebBaseLoader(first_user_message)
+
+            openai_embedding = OpenAIEmbeddings(
+                openai_api_base=DIAL_URL,
+                openai_api_key=request.api_key,
+                openai_api_version="2023-03-15-preview",
+                openai_api_type="azure",
+                max_retries=15,
+                headers=(
+                    {} if not request.jwt else {"Authorization": request.jwt}
+                ),
+            )
+
+            embeddings = CacheBackedEmbeddings.from_bytes_store(
+                openai_embedding,
+                embedding_store,
+                namespace=openai_embedding.model,
+            )
+
+            if len(request.messages) == 1:
+                # Create the download stage to show to the user the active process.
+                # After the loading is complete, the stage will auto finished.
+                with choice.create_stage("Download the resource"):
```

Comment:
`Downloading the resource` as a continious process looks better as a stage name. Same for other stages.

`Download the resource` sounds like a command: `You! Download the resource`.